.github/bin/build-matrix-from-impacted.py

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
+
+import argparse
+import yaml
+import json
+import logging
+import sys
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Filter test matrix modules using list of impacted modules."
+    )
+    parser.add_argument(
+        "-m",
+        "--matrix",
+        type=argparse.FileType("r"),
+        default=".github/test-matrix.yaml",
+        help="A YAML file with the test matrix",
+    )
+    parser.add_argument(
+        "-i",
+        "--impacted",
+        type=argparse.FileType("r"),
+        default="gib-impacted.log",
+        help="List of impacted modules, one per line",
+    )
+    parser.add_argument(
+        "-o",
+        "--output",
+        type=argparse.FileType("w"),
+        default=sys.stdout,
+        help="Filename to write JSON output to",
+    )
+    parser.add_argument(
+        "-d",
+        "--debug",
+        action="store_const",
+        dest="loglevel",
+        const=logging.DEBUG,
+        default=logging.WARNING,
+        help="Print lots of debugging statements",
+    )
+
+    args = parser.parse_args()
+    logging.basicConfig(level=args.loglevel)
+    build(args.matrix, args.impacted, args.output)
+
+
+def build(matrix_file, impacted_file, output_file):
+    matrix = yaml.load(matrix_file, Loader=yaml.Loader)
+    impacted = list(filter(None, [line.rstrip() for line in impacted_file.readlines()]))
+    logging.debug("Read matrix: %s", matrix)
+    logging.debug("Read impacted: %s", impacted)
+    matrix["include"] = [
+        {
+            # concatenate because matrix values should be primitives
+            "modules": ",".join(item["modules"]),
+            "options": item.get("options", ""),
+        }
+        for item in matrix.get("include", [])
+        if any(module in impacted for module in item.get("modules", []))
```

Comment:
Reads very weridly. Is this supposed to be an empty loop?

Reply:
No, there are no empty loops here. This is regular list comprehension in Python. We're:
* rebuilding the `matrix["include"]` list, collapsing `modules` key from a list to a string
* by iterating over all values
* only including entries with any module in the impacted modules list - we're using another list comprehension inside `any()`

I can avoid using list comprehension if you'd like me to.

Reply:
Oh, now I notice that the entire expression is within `matrix["include"] = [...]`.

I'd prefer the more readable choice (I don't know how the alternative looks like - it's fine if list comprehension is more readable).

Reply:
An alternative is a regular loop that appends to a list:
```
    matrix["include"] = []
    for item in matrix.get("include", []):
        if not any(module in impacted for module in item.get("modules", [])):
            continue
        matrix["include"].append({
            # concatenate because matrix values should be primitives
            "modules": ",".join(item["modules"]),
            "options": item.get("options", ""),
        })
```

Reply:
Hmm, isn't worse. I prefer the loop - less confusing.

Reply:
Done!